File: src/main/java/org/bk/graphql/entity/BookEntity.java
package org.bk.graphql.entity;

import org.bk.graphql.domain.Book;
import org.bk.graphql.domain.BookId;
import org.bk.graphql.domain.ImmutableBook;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Version;
import org.springframework.data.relational.core.mapping.Table;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

@Table("books")
public record BookEntity(
        @Id UUID id,
        String name,
        int pageCount,
        Instant createdAt,
        Instant updatedAt,
        @Version int version
) {
    public static BookEntity fromModel(Book book) {
        return new BookEntity(book.id().id(), book.name(), book.pageCount(), book.createdAt(), book.updatedAt(), book.version());
    }

    public Book toModel() {
        return ImmutableBook.builder()
                .id(BookId.of(id))
                .name(name)
                .pageCount(pageCount)
                //Placeholder..will be filled in later.
                .authors(List.of())
                .createdAt(createdAt)
                .updatedAt(updatedAt)
                .version(version)
                .build();
    }
}

File: src/main/kotlin/org/bk/graphql/entity/BookEntity.kt
package org.bk.graphql.entity

import org.bk.graphql.domain.AuthorId
import org.bk.graphql.domain.Book
import org.bk.graphql.domain.BookId
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Version
import org.springframework.data.relational.core.mapping.MappedCollection
import org.springframework.data.relational.core.mapping.Table

@Table("books")
data class BookEntity(
    @Id val id: String,
    val name: String,
    val pageCount: Int,
    @MappedCollection(idColumn = "book_id")
    val authors: Set<AuthorRef>,
    @Version val version: Int = 0
) {
    fun toModel() = Book(
        id = BookId(id),
        name = name,
        pageCount = pageCount,
        authors = authors.stream().map { author -> AuthorId(author.author.id!!) }.toList(),
        version = version
        )
}

